src/server/lock_manager/deadlock.rs

```diff
@@ -390,20 +383,99 @@ impl Scheduler {
+        // of the deadlock detector stepping down.
+        //
+        // If the peers of a region is not empty, the region info is complete.
+        !region.get_peers().is_empty()
+            && region.get_start_key() <= LEADER_KEY
+            && (region.get_end_key().is_empty() || LEADER_KEY < region.get_end_key())
+    }
+
+    pub(crate) fn new(scheduler: Scheduler) -> Self {
+        Self {
+            leader_region_id: Arc::new(Mutex::new(INVALID_ID)),
+            scheduler,
+        }
+    }
+
+    pub(crate) fn register(self, host: &mut CoprocessorHost) {
+        host.registry
+            .register_role_observer(1, Box::new(self.clone()));
+        host.registry
+            .register_region_change_observer(1, Box::new(self));
+    }
+}
+
+impl Coprocessor for RoleChangeNotifier {}
+
+impl RoleObserver for RoleChangeNotifier {
     fn on_role_change(&self, ctx: &mut ObserverContext<'_>, role: StateRole) {
-        if is_leader_region(ctx.region()) {
-            self.change_role(role.into());
+        let region = ctx.region();
+        // A region is created first, so the leader region id must be valid.
+        if Self::is_leader_region(region)
+            && *self.leader_region_id.lock().unwrap() == region.get_id()
+        {
+            self.scheduler.change_role(role.into());
+        }
+    }
+}
+
+impl RegionChangeObserver for RoleChangeNotifier {
+    fn on_region_changed(
+        &self,
+        ctx: &mut ObserverContext<'_>,
+        event: RegionChangeEvent,
+        role: StateRole,
+    ) {
+        let region = ctx.region();
+        if Self::is_leader_region(region) {
+            match event {
+                RegionChangeEvent::Create | RegionChangeEvent::Update => {
+                    *self.leader_region_id.lock().unwrap() = region.get_id();
+                    self.scheduler.change_role(role.into());
+                }
+                RegionChangeEvent::Destroy => {
+                    // When one region is merged to target region, it will be destroyed.
+                    // If the leader region is merged to the target region and the node
+                    // is also the leader of the target region, the RoleChangeNotifier will
+                    // receive one RegionChangeEvent::Update of the target region and one
+                    // RegionChangeEvent::Destroy of the leader region. To prevent the
+                    // destroy event misleading the leader stepping down, it saves the
+                    // valid leader region id and only when the id equals to the destroyed
+                    // region id, it sends a ChangeRole(Follower) task to the deadlock detector.
+                    let mut leader_region_id = self.leader_region_id.lock().unwrap();
+                    if *leader_region_id == region.get_id() {
+                        *leader_region_id = INVALID_ID;
+                        self.scheduler.change_role(Role::Follower);
```

Comment:
Is it possible that when a merge happens, a `Destroy` happens first and an `Update` follows? May there be no leader between the two message?

Reply:
Yes. The `Update` will be coming soon since the `Destroy` happens.